sway/commands/resize.c

```diff
@@ -0,0 +1,297 @@
+#include <errno.h>
+#include <math.h>
+#include <stdbool.h>
+#include <stdlib.h>
+#include <string.h>
+#include <strings.h>
+#include <wlr/util/log.h>
+#include "sway/commands.h"
+#include "sway/tree/layout.h"
+#include "log.h"
+
+static const int min_sane_w = 100, min_sane_h = 60;
+
+enum resize_unit {
+	RESIZE_UNIT_PX,
+	RESIZE_UNIT_PPT,
+	RESIZE_UNIT_DEFAULT,
+	RESIZE_UNIT_INVALID,
+};
+
+enum resize_axis {
+	RESIZE_AXIS_HORIZONTAL,
+	RESIZE_AXIS_VERTICAL,
+	RESIZE_AXIS_INVALID,
+};
+
+static enum resize_unit parse_resize_unit(const char *unit) {
+	if (strcasecmp(unit, "px") == 0) {
+		return RESIZE_UNIT_PX;
+	}
+	if (strcasecmp(unit, "ppt") == 0) {
+		return RESIZE_UNIT_PPT;
+	}
+	if (strcasecmp(unit, "default") == 0) {
+		return RESIZE_UNIT_DEFAULT;
+	}
+	return RESIZE_UNIT_INVALID;
+}
+
+static enum resize_axis parse_resize_axis(const char *axis) {
+	if (strcasecmp(axis, "width") == 0 || strcasecmp(axis, "horizontal") == 0) {
+		return RESIZE_AXIS_HORIZONTAL;
+	}
+	if (strcasecmp(axis, "height") == 0 || strcasecmp(axis, "vertical") == 0) {
+		return RESIZE_AXIS_VERTICAL;
+	}
+	return RESIZE_AXIS_INVALID;
+}
+
+static int parallel_coord(struct sway_container *c, enum resize_axis a) {
+	return a == RESIZE_AXIS_HORIZONTAL ? c->x : c->y;
+}
+
+static int parallel_size(struct sway_container *c, enum resize_axis a) {
+	return a == RESIZE_AXIS_HORIZONTAL ? c->width : c->height;
+}
+
+static void resize_tiled(int amount, enum resize_axis axis) {
+	struct sway_container *parent = config->handler_context.current_container;
+	struct sway_container *focused = parent;
+	struct sway_container *sibling;
+	if (!parent) {
+		return;
+	}
+
+	enum sway_container_layout parallel_layout =
+		axis == RESIZE_AXIS_HORIZONTAL ? L_HORIZ : L_VERT;
+	int minor_weight = 0;
+	int major_weight = 0;
+	while (parent->parent) {
+		struct sway_container *next = parent->parent;
+		if (next->layout == parallel_layout) {
+			for (int i = 0; i < next->children->length; i++) {
+				sibling = next->children->items[i];
+
+				int sibling_pos = parallel_coord(sibling, axis);
+				int focused_pos = parallel_coord(focused, axis);
+				int parent_pos = parallel_coord(parent, axis);
+
+				if (sibling_pos != focused_pos) {
+					if (sibling_pos < parent_pos) {
+						minor_weight++;
+					} else if (sibling_pos > parent_pos) {
+						major_weight++;
+					}
+				}
+			}
+			if (major_weight || minor_weight) {
+				break;
+			}
+		}
+		parent = next;
+	}
+
+	if (parent->type == C_ROOT) {
+		return;
+	}
+
+	wlr_log(L_DEBUG,
+			"Found the proper parent: %p. It has %d l conts, and %d r conts",
+			parent->parent, minor_weight, major_weight);
+
+	int min_sane = axis == RESIZE_AXIS_HORIZONTAL ? min_sane_w : min_sane_h;
+
+	//TODO: Ensure rounding is done in such a way that there are NO pixel leaks
+	// ^ ?????
+
+	for (int i = 0; i < parent->parent->children->length; i++) {
```

Comment:
`parent->parent` can be NULL.

Reply:
cannot

Reply:
The loop above breaks if `parent->parent == NULL`.

Reply:
Which it can't be. The only case where that's true is containers detached from the main tree or the root container, and the former can't happen here and the latter is explicitly checked for.

Reply:
Then check for the container type in the loop condition instead of checking for `parent->parent`?

Reply:
I don't think this needs to change.